ir/ir.rs

```diff
@@ -33,68 +33,66 @@ impl ItemsBuilder {
     /// Construct a new builder, with the given size.
     pub fn new(size: u32) -> ItemsBuilder {
         ItemsBuilder {
-            id_counter: 0,
             size,
             parsed: Default::default(),
             items: Default::default(),
+            data:  Default::default(),
             edges: Default::default(),
             roots: Default::default(),
         }
     }
 
-    fn next_id(&mut self) -> Id {
-        let id = Id(self.id_counter);
-        self.id_counter += 1;
-        id
-    }
-
     /// Add the given item to to the graph and return the `Id` that it was
     /// assigned.
-    pub fn add_item<T>(&mut self, key: &T, mut item: Item) -> Id {
-        let id = self.next_id();
-
-        item.id = id;
+    pub fn add_item(&mut self, item: Item) -> Id {
+        let id = item.id;
         self.items.insert(id, item);
 
-        let old_value = self.parsed.insert(key as *const T as *const (), id);
-        assert!(
-            old_value.is_none(),
-            "should not parse the same key into multiple items"
-        );
-
+        let old_value = self.parsed.insert(id);
+        assert!(old_value, "should not parse the same key into multiple items");
+        
         id
     }
 
     /// Add the given item to the graph as a root and return the `Id` that it
     /// was assigned.
-    pub fn add_root<T>(&mut self, key: &T, item: Item) -> Id {
-        let id = self.add_item(key, item);
+    pub fn add_root(&mut self, item: Item) -> Id {
+        let id = self.add_item(item);
         self.roots.insert(id);
         id
     }
 
     /// Add an edge between the given keys that have already been parsed into
     /// items.
-    pub fn add_edge<K, J>(&mut self, from: &K, to: &J) {
-        let from_id = self.id_for_key(from);
-        let to_id = self.id_for_key(to);
+    pub fn add_edge(&mut self, from: Id, to: Id) {
         self.edges
-            .entry(from_id)
+            .entry(from)
             .or_insert(BTreeSet::new())
-            .insert(to_id);
+            .insert(to);
     }
 
-    /// Get the id for the item we parsed from the given key.
-    pub fn id_for_key<T>(&self, key: &T) -> Id {
-        let key = key as *const T as *const ();
-        self.parsed[&key]
+    /// Add a range of static data and the `Id` that defines it.
+    pub fn link_data(&mut self, offset: i64, len: usize, id: Id) {
+        if offset >= 0 && offset <= u32::MAX as i64 && offset as usize + len < u32::MAX as usize {
+            self.data.insert(offset as u32, (id, len as u32));
+        }
+    }
+
+    /// locate the data section defining memory at the given offset
+    pub fn get_data(&self, offset: u32) -> Option<Id> {
+        self.data.range(offset..).next().and_then(|(start, &(id, len))| {
+            if offset < start + len {
+                Some(id)
+            } else {
+                None
+            }
+        })                                         
```

Comment:
Nitpick: don't include trailing white space. Did you run `cargo fmt --all`? I don't think there would be trailing whitespace if so.

https://github.com/fitzgen/svelte/blob/master/CONTRIBUTING.md#automatic-code-formatting